### PR TITLE
Reg support radio test

### PR DESCRIPTION
**Pull request by karun2796**
(1) drivers: wifi: Setting default country code
     [SHEL-1346] Setting default regulatory country code at build time.
(2) samples: wifi: radio_test: Add regulatory support
     [SHEL-1468] Add regulatory support.
(3) drivers: wifi: Update RPU patch to rev#116
    [SHEL-1339]: Regulatory support in radio test mode.
    [SHEL-1346]: Configuring default country code during init command in
                 regular mode.
    [SHEL-1446]: Non-consecutive Packet Number attack test fix.
    [SHEL-1200]: Place RF in Standby mode after completing PHY_OPEN
                 operation.
    [SHEL-1425]: Adding commands to configure Extended power save
                 & listen interval.
    [SHEL-1441]: Faking A-MSDUs as EAPOL after a secure connection test fix.
    [SHEl-1451]: Multiple plaintext fragment attack test fix.
    [SHEL-1415]: Listen interval based power save.

### CODE REVIEW

File: samples/wifi/radio_test/radio_test_subcommands.rst [hunk 1]
```diff
@@ -237,6 +237,19 @@ Wi-Fi radio test subcommands
      - 15
      - Configuration
      - Contention window for transmitted packets.
+   * - reg_domain
+     - <country code> – Desired country code(exp: NO, US, UK, IN).
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/radio_test_subcommands.rst:241]: I am assuming by `exp` you mean example. Please use `e.g.` instead.

File: samples/wifi/radio_test/radio_test_subcommands.rst [hunk 2]
```diff
@@ -237,6 +237,19 @@ Wi-Fi radio test subcommands
...
+     - 00(World regulatory)
+     - Action
+     - Configure WLAN regulatory domain country code.
+   * - bypass_reg_domain
+     - | 0: Use reg_domain
+       | 1: Do not use reg_domain
+     - 0
+     - Configuration
+     - Configure WLAN to bypass current regulatory domain in tx test.
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/radio_test_subcommands.rst:250]: use TX instead of tx

File: samples/wifi/radio_test/sample_description.rst [hunk 3]
```diff
@@ -559,6 +560,50 @@ Testing
               [00:24:25.203,002] <inf> otp_prog: Finished Writing OTP params
 
          See :ref:`wifi_radio_ficr_prog_subcmds` for a list of available subcommands.
+         
+         * To set regulatory domain with the following configuration:
+
+           * regulatoy domain: US
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/sample_description.rst:566]: regulatory spelling

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 4]
```diff
@@ -1396,6 +1396,33 @@ enum wifi_nrf_status nrf_wifi_fmac_rf_test_compute_xo(struct wifi_nrf_fmac_dev_c
...
+		wifi_nrf_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+				      "%s: Invalid parameters\n",
+				      __func__);
+		goto out;
+	}
+
+	status = umac_cmd_prog_reg_domain(fmac_dev_ctx,
+					  country_code);
+
+
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:1415]: Unnecessary extra blank line.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 5]
```diff
@@ -1396,6 +1396,33 @@ enum wifi_nrf_status nrf_wifi_fmac_rf_test_compute_xo(struct wifi_nrf_fmac_dev_c
...
+	if (status != WIFI_NRF_STATUS_SUCCESS) {
+		wifi_nrf_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+				      "%s: Unable to program regulatory \n",
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:1418]: Remove space before `\n`

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 6]
```diff
@@ -1990,6 +2001,81 @@ static int nrf_wifi_radio_test_wlan_switch_ctrl(const struct shell *shell,
...
+	if (status != WIFI_NRF_STATUS_SUCCESS) {
+		shell_fprintf(shell,
+			      SHELL_ERROR,
+			      "Regulatory programming failed\n");
+		goto out;
+	}
+
+	ret = 0;
+out:
+
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2046]: Unnecessary blank line.

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 7]
```diff
@@ -1990,6 +2001,81 @@ static int nrf_wifi_radio_test_wlan_switch_ctrl(const struct shell *shell,
...
+					      size_t argc,
+					      const char *argv[])
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	int ret = -ENOEXEC;
+
+	if (strlen(argv[1]) != 2) {
+		shell_fprintf(shell, SHELL_WARNING,
+			"Invalid reg domain: Length should be two letters/digits\n");
+			return -ENOEXEC;
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2015]: Use `goto out;`

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 8]
```diff
@@ -1990,6 +2001,81 @@ static int nrf_wifi_radio_test_wlan_switch_ctrl(const struct shell *shell,
...
+	}
+
+	/* Two letter country code with special case of 00 for WORLD */
+	if (((argv[1][0] < 'A' || argv[1][0] > 'Z') ||
+	     (argv[1][1] < 'A' || argv[1][1] > 'Z')) &&
+	     (argv[1][0] != '0' || argv[1][1] != '0')) {
+		shell_fprintf(shell, SHELL_WARNING, "Invalid reg domain %c%c\n", argv[1][0],
+			      argv[2][1]);
+		return -ENOEXEC;
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2024]: Use `goto out;`

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 9]
```diff
@@ -1990,6 +2001,81 @@ static int nrf_wifi_radio_test_wlan_switch_ctrl(const struct shell *shell,
...
+	}
+
+	ctx->conf_params.country_code[0] = argv[1][0];
+	ctx->conf_params.country_code[1] = argv[1][1];
+
+	if (!check_test_in_prog(shell)) {
+		return -ENOEXEC;
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2031]: Use `goto out;`

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: samples/wifi/radio_test/sample_description.rst [hunk 10]
```diff
@@ -560,6 +561,50 @@ Testing
...
+               wifi_radio_test show_config
+               reg_domain = US
+
+         .. note::
+
+            The default regulatory domain is 00(World regulatory).
+
+         * To bypass regulatory domain set use the following configuration:
+
+           * bypass_reg_domain: 1
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:587]:
```suggestion
         * To bypass regulatory domain, set ``bypass_reg_domain`` to ``1`` using the following command:
```

File: samples/wifi/radio_test/sample_description.rst [hunk 11]
```diff
@@ -560,6 +561,50 @@ Testing
...
+
+           Execute the following sequence of commands:
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:589]:
```suggestion
```

File: samples/wifi/radio_test/sample_description.rst [hunk 12]
```diff
@@ -560,6 +561,50 @@ Testing
 
          See :ref:`wifi_radio_ficr_prog_subcmds` for a list of available subcommands.
 
+         * To set regulatory domain with the following configuration:
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:564]: Line 562 "See :ref:`wifi_radio_ficr_prog_subcmds` for a list of available subcommands." can be moved after line 605.

**Review by richabp — CHANGES_REQUESTED** [review]

File: samples/wifi/radio_test/sample_description.rst [hunk 13]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
+              wifi_radio_test bypass_reg_domain 1
+
+           The sample shows the following output:
+
+           .. code-block:: console
+
+               wifi_radio_test show_config
+               reg_domain = US
+               bypass_reg_domain = 1
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:601]: Lines 585-601 need not to be in Note.

File: samples/wifi/radio_test/sample_description.rst [hunk 14]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
+				reg_domain = US
+
+         .. note::
+
+            * The default regulatory domain is 00 (World regulatory).
+
+		 * To bypass regulatory domain, set ``bypass_reg_domain`` to ``1`` using the following command:
+
+			* bypass_reg_domain: 1
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:587]:
```suggestion

```

File: samples/wifi/radio_test/sample_description.rst [hunk 15]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
+           .. code-block:: console
+
+				wifi_radio_test show_config
+
+				reg_domain = US
+
+         .. note::
+
+            * The default regulatory domain is 00 (World regulatory).
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:583]:
```suggestion
            The default regulatory domain is 00 (World regulatory).
```

File: samples/wifi/radio_test/sample_description.rst [hunk 16]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
+           .. code-block:: console
+
+               wifi_radio_test show_config
+               reg_domain = US
+               bypass_reg_domain = 1
+
+         .. note::
+
+            * Bypass regulatory domain is false by default.
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:605]:
```suggestion
            Bypass regulatory domain is false by default.
```

File: samples/wifi/radio_test/sample_description.rst [hunk 17]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
+            See :ref:`wifi_radio_ficr_prog_subcmds` for a list of available subcommands.
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:607]: Line 607: It should be out of the Note.

File: samples/wifi/radio_test/sample_description.rst [hunk 18]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
+         .. note::
+
+            * The default regulatory domain is 00 (World regulatory).
+
+		 * To bypass regulatory domain, set ``bypass_reg_domain`` to ``1`` using the following command:
+
+			* bypass_reg_domain: 1
+
+           Execute the following sequence of command:
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:589]:
```suggestion
    
```

File: samples/wifi/radio_test/sample_description.rst [hunk 19]
```diff
@@ -558,7 +559,53 @@ Testing
...
+
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:590]:
```suggestion
```

**Review by richabp — CHANGES_REQUESTED** [review]

File: samples/wifi/radio_test/sample_description.rst (hunk 19 above)
Comment (karun2796) [reply samples/wifi/radio_test/sample_description.rst:590]: This is probably needed.  (In line 590 a blank line is needed)

File: samples/wifi/radio_test/radio_test_subcommands.rst [hunk 20]
```diff
@@ -237,6 +237,19 @@ Wi-Fi radio test subcommands
      - 15
      - Configuration
      - Contention window for transmitted packets.
+   * - reg_domain
+     - <country code> – Desired country code(for example: NO, US, UK, IN).
```
Comment (sachinthegreen) [inline samples/wifi/radio_test/radio_test_subcommands.rst:241]: It was suggested that `for example` be replaced by `e.g.`

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

**Review by sachinthegreen — APPROVED** [review]

**Review by richabp — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 21]
```diff
@@ -225,6 +225,21 @@ enum wifi_nrf_status nrf_wifi_fmac_set_xo_val(struct wifi_nrf_fmac_dev_ctx *fmac
...
+ * @fmac_dev_ctx: Pointer to the UMAC IF context for a RPU WLAN device.
+ *
+ * This function is used to send a command to RPU to
+ * configure regulatory country code.
+ *
+ * Returns: Status
+ *              Pass : %WIFI_NRF_STATUS_SUCCESS
+ *              Error: %WIFI_NRF_STATUS_FAIL
+ */
+enum wifi_nrf_status nrf_wifi_fmac_set_reg_domain(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:241]: We already have a API for this, just need to move out of `#ifndef RADIO_TEST`

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Review by bama-nordic — APPROVED** [review]

File: samples/wifi/radio_test/sample_description.rst [hunk 22]
```diff
@@ -558,8 +559,49 @@ Testing
...
+           .. code-block:: console
+
+              wifi_radio_test show_config
+              reg_domain = US
+
+         .. note::
+
+            The default regulatory domain is ``00`` (World regulatory).
+
+         * To bypass regulatory domain, set ``bypass_reg_domain`` to ``1`` using the following command:
```
Comment (krish2718) [inline samples/wifi/radio_test/sample_description.rst:584]: Exact meaning of `byapss` needs some doc, e.g., which rules it follow? or just accepts any channel or power etc?

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 23]
```diff
@@ -2262,6 +2350,20 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
...
+		      NULL,
+		      "Configure WLAN regulatory domain country code",
+		      nrf_wifi_radio_test_set_reg_domain,
+		      2,
+		      0),
+	SHELL_CMD_ARG(bypass_reg_domain,
+		      NULL,
+		      "Configure WLAN to by-pass regulatory\n"
+		      "0 - Use reg_domain\n"
+		      "1 - Do not use reg_domain						",
```
Comment (krish2718) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2363]: In bypass there is not reg_domain, so, `do not use reg_domain` is quite right.

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 24]
```diff
@@ -2262,6 +2350,20 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
...
 		      0),
+	SHELL_CMD_ARG(reg_domain,
+		      NULL,
+		      "Configure WLAN regulatory domain country code",
+		      nrf_wifi_radio_test_set_reg_domain,
+		      2,
+		      0),
+	SHELL_CMD_ARG(bypass_reg_domain,
+		      NULL,
+		      "Configure WLAN to by-pass regulatory\n"
```
Comment (krish2718) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2361]: bypass

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c [hunk 25]
```diff
@@ -1990,6 +2001,83 @@ static int nrf_wifi_radio_test_wlan_switch_ctrl(const struct shell *shell,
...
+	return ret;
+}
+
+
+static int nrf_wifi_radio_test_set_bypass_reg(const struct shell *shell,
+					      size_t argc,
+					      const char *argv[])
+{
+	char *ptr = NULL;
+	unsigned long val = 0;
```
Comment (krish2718) [inline samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2058]: bool should suffice

File: drivers/wifi/nrf700x/Kconfig.nrf700x [hunk 26]
```diff
@@ -27,6 +27,10 @@ if WIFI_NRF700X
 config NRF700X_TCP_IP_CHECKSUM_OFFLOAD
 	bool "Enable TCP/IP checksum offload"
 
+config NRF700X_REG_DOMAIN
+	string "The ISO/IEC alpha2 country code for the country in which this device is currently operating."
+	default "00"
```
Comment (krish2718) [inline drivers/wifi/nrf700x/Kconfig.nrf700x:32]: as 00 is special value, may be add a comment here.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: samples/wifi/radio_test/sample_description.rst (hunk 22 above)
Comment (ajayparida) [reply samples/wifi/radio_test/sample_description.rst:584]: updated with description

File: samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c (hunk 25 above)
Comment (ajayparida) [reply samples/wifi/radio_test/src/nrf_wifi_radio_test_shell.c:2058]: This is not required. We have to handle user provided values.

**Review by udaynordic — APPROVED** [review]

File: samples/wifi/radio_test/sample_description.rst [hunk 27]
```diff
@@ -558,8 +559,55 @@ Testing
...
+               wifi_radio_test show_config
+               reg_domain = US
+               bypass_reg_domain = 1
+
+         .. note::
+
+            Bypass regulatory domain is false by default.
+
+            If 'bypass_reg_domain = 0', then TX power of the channel will be configured to the minimum value of
+            the user configured TX power value and max power supported in the configured regulatory domain.
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:603]:
```suggestion
            If ``bypass_reg_domain`` is ``0``, then TX power of the channel will be configured to the minimum value of the user configured TX power value and maximum power supported in the configured regulatory domain.
```

File: samples/wifi/radio_test/sample_description.rst [hunk 28]
```diff
@@ -558,8 +559,55 @@ Testing
...
+
+            If 'bypass_reg_domain = 1', then user configured TX power value will be set overriding current
+            configured regulatory domain max TX power for the channel.
```
Comment (richabp) [inline samples/wifi/radio_test/sample_description.rst:606]:
```suggestion
            If ``bypass_reg_domain`` is ``1``, then user configured TX power value will be set overriding current configured regulatory domain maximum TX power for the channel.
```

**Review by richabp — CHANGES_REQUESTED** [review]

**Review by richabp — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 29]
```diff
@@ -4274,6 +4220,55 @@ enum wifi_nrf_status wifi_nrf_fmac_rf_params_get(struct wifi_nrf_fmac_dev_ctx *f
...
+		goto out;
+	}
+
+	set_reg_cmd->umac_hdr.cmd_evnt = NRF_WIFI_UMAC_CMD_REQ_SET_REG;
+	set_reg_cmd->umac_hdr.ids.valid_fields = 0;
+
+	wifi_nrf_osal_mem_cpy(fmac_dev_ctx->fpriv->opriv,
+						  set_reg_cmd->nrf_wifi_alpha2,
+						  reg_info->alpha2,
+						  NRF_WIFI_COUNTRY_CODE_LEN);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:4252]: Indentation seems to be off.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 30]
```diff
@@ -4274,6 +4220,55 @@ enum wifi_nrf_status wifi_nrf_fmac_rf_params_get(struct wifi_nrf_fmac_dev_ctx *f
...
+
+	set_reg_cmd->valid_fields = NRF_WIFI_CMD_REQ_SET_REG_ALPHA2_VALID;
+
+	/* New feature in rev B patch */
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:4256]: Is this comment needed?

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c (hunk 30 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:4256]: This was pre-existing comment. 
@krish2718 Please update.

**Comment by krish2718** [comment]
I guess https://github.com/nrfconnect/sdk-nrf/pull/10790/commits supersedes this, so, can this be closed?

**Comment by ajayparida** [comment]
Closing this PR, as the changes are part of https://github.com/nrfconnect/sdk-nrf/pull/10790/commits

**Pull request closed without merging**